File: app/src/io/reportClient.ts
/**
 * reportClient — submits a Help ▸ Report Bug report to the intake service
 * (docs/design/report-bug.md §8, the wire contract this must match exactly;
 * `workers/bug-intake` is the other side, built separately).
 *
 * A report is the gzip of the bundle `reportBundle.ts` builds, uploaded in
 * pieces of `PIECE_BYTES`: a start request carrying the first piece and the
 * total length, each following piece in order, then a commit. Each request
 * stays small because the intake service runs where a request gets 10 ms of
 * CPU (§4). A request that fails with a network error or a 5xx is retried
 * once.
 *
 * Desktop runs that whole upload in the Rust `report_submit` command
 * (report_client.rs) — the same "no arbitrary-URL capability in the
 * webview" posture as `relayClient.ts` — and always can, in every server
 * mode (design §5: a bug report is for Hew's developer, not a self-hosted
 * homelab). Web uploads same-origin under `/report/`, which only makes
 * sense — and is only attempted — when the page's own origin actually
 * serves that route: the production `app.hew3d.com` deployment, or a local
 * `vite dev` with `HEW_REPORT_PROXY` forwarding it to `wrangler dev` (see
 * `vite.config.ts`). Every other web build (self-hosted, a plain dev
 * server) has nowhere to send to, so `canSend()` is false there and
 * `ReportBugDialog` hides Send report entirely rather than let a submit
 * attempt fail.
 */

import { isTauri } from './fileHost'

/** Mirrors `ReportErrorKind` in report_client.rs (serde camelCase), plus
 *  `offline` for a web `fetch` that never reached the network at all —
 *  the desktop command's transport failures already collapse to
 *  `unreachable`/`tls`, but a browser `fetch` throwing a bare `TypeError`
 *  carries no such distinction from Rust to preserve. */
export type ReportErrorKind =
  | 'invalid'
  | 'tooLarge'
  | 'rateLimited'
  | 'full'
  | 'unreachable'
  | 'tls'
  | 'status'
  | 'offline'

export class ReportError extends Error {
  readonly kind: ReportErrorKind
  readonly status: number | undefined

  constructor(kind: ReportErrorKind, message: string, status?: number) {
    super(message)
    this.name = 'ReportError'
    this.kind = kind
    this.status = status
  }
}

const KINDS: ReadonlySet<string> = new Set<ReportErrorKind>([
  'invalid',
  'tooLarge',
  'rateLimited',
  'full',
  'unreachable',
  'tls',
  'status',
  'offline',
])

/** Turns whatever the Tauri `report_submit` invoke rejected with into a
 *  `ReportError`. Mirrors `relayClient.ts`'s `toRelayError`. */
function toReportError(err: unknown): ReportError {
  if (err instanceof ReportError) return err
  if (typeof err === 'object' && err !== null && 'kind' in err) {
    const kind = (err as { kind: unknown }).kind
    const message = (err as { message?: unknown }).message
    const status = (err as { status?: unknown }).status
    if (typeof kind === 'string' && KINDS.has(kind)) {
      return new ReportError(
        kind as ReportErrorKind,
        typeof message === 'string' ? message : kind,
        typeof status === 'number' ? status : undefined,
      )
    }
  }
  return new ReportError('unreachable', err instanceof Error ? err.message : String(err))
}

/** The production intake host — the one web origin `/report/` uploads are
 *  ever attempted from outside a proxied dev server. */
const CLOUD_APP_ORIGIN = 'https://app.hew3d.com'

/** Set by `vite.config.ts`'s `define` when `HEW_REPORT_PROXY` was configured
 *  for this dev server — see that file for why this needs to be a
 *  build-time flag rather than a runtime env read (the browser has no
 *  `process.env`). Undefined (falsy) outside a Vite build (e.g. plain
 *  `tsc`/vitest), matching the `__HEW_VERSION__` fallback convention. */
declare const __HEW_REPORT_DEV_PROXY__: boolean | undefined

/**
 * Whether this build/runtime can actually send a report anywhere. Desktop:
 * always (every server mode posts to the cloud intake service). Web: only
 * the production app origin, or a dev server with the proxy configured.
 * Everything else (a self-hosted web build, `vite dev` with no proxy) has
 * no route to the intake service, so the dialog must not offer Send report.
 */
export function canSend(): boolean {
  if (isTauri) return true
  if (typeof location !== 'undefined' && location.origin === CLOUD_APP_ORIGIN) return true
  return typeof __HEW_REPORT_DEV_PROXY__ !== 'undefined' && __HEW_REPORT_DEV_PROXY__
}

export interface ReportSubmitResult {
  id: string
}

/** How much of the compressed report has been uploaded. */
export interface ReportProgress {
  sentBytes: number
  totalBytes: number
}

/** The upload piece size, one storage row on the intake service (§8). */
export const PIECE_BYTES = 1_900_000

/** One request carries at most one piece; two minutes covers that on a slow
 *  uplink and still ends a stalled connection. */
const REQUEST_TIMEOUT_MS = 120_000

/** Maps one of §8's JSON error bodies + HTTP status onto a typed kind —
 *  the web-`fetch` twin of report_client.rs's `ReportError::from_status`.
 *  Falls back to `status` for anything that doesn't match the documented
 *  shape, which includes a piece's 403/404/409. */
function reportErrorFromResponse(status: number, body: unknown): ReportError {
  const error = typeof body === 'object' && body !== null ? (body as { error?: unknown }).error : undefined
  switch (`${status} ${typeof error === 'string' ? error : ''}`) {
    case '400 invalid':
      return new ReportError('invalid', 'the report was rejected as invalid', status)
    case '413 too-large':
      return new ReportError('tooLarge', 'the report is too large to send', status)
    case '429 rate-limited':
      return new ReportError('rateLimited', 'too many reports sent recently', status)
    case '507 full':
      return new ReportError('full', 'the report service is temporarily full', status)
    default:
      if (status >= 500) return new ReportError('unreachable', 'the report service is unavailable', status)
      return new ReportError('status', `unexpected status ${status}`, status)
  }
}

/** Desktop path: the gzip bytes ride the invoke as a RAW body (the same
 *  convention `relayClient.ts`'s `relayPut` uses), never a JSON array — a
 *  90 MiB upload must not be serialized as tens of millions of numbers.
 *  Rust runs the piece-by-piece upload and reports progress as
 *  `report-progress` events. */
async function submitDesktop(
  gzipBytes: Uint8Array,
  onProgress?: (progress: ReportProgress) => void,
): Promise<ReportSubmitResult> {
  const { invoke } = await import('@tauri-apps/api/core')
  let unlisten: (() => void) | undefined
  if (onProgress !== undefined) {
    try {
      const { listen } = await import('@tauri-apps/api/event')
      unlisten = await listen<ReportProgress>('report-progress', (event) => onProgress(event.payload))
    } catch {
      // Progress is cosmetic; the upload doesn't depend on it.
    }
  }
  try {
    return await invoke<ReportSubmitResult>('report_submit', gzipBytes as never)
  } catch (err) {
    throw toReportError(err)
  } finally {
    unlisten?.()
  }
}

/** One `fetch` with the per-request timeout. A thrown error (offline, DNS,
 *  aborted by the timeout) comes back as `null` instead of a response. */
async function fetchOnce(url: string, init: RequestInit): Promise<Response | null> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS)
  try {
    return await fetch(url, { ...init, signal: controller.signal })
  } catch {
    return null
  } finally {
    clearTimeout(timer)
  }
}

/** §8's retry rule: a network error or a 5xx is retried once. Every step
 *  tolerates a repeat: the service accepts a retry of the last piece, and a
 *  commit repeated with the upload's token answers 201 again. */
async function fetchWithRetry(url: string, init: RequestInit): Promise<Response> {
  const first = await fetchOnce(url, init)
  if (first !== null && first.status < 500) return first
  const second = await fetchOnce(url, init)
  if (second === null) throw new ReportError('offline', 'could not reach the report service')
  return second
}

async function readJson(response: Response): Promise<unknown> {
  try {
    return await response.json()
  } catch {
    return null
  }
}

/** One piece of the upload as a request body. `gzip()` allocates a plain
 *  ArrayBuffer, so the view is safe to hand to `fetch` as-is. */
function pieceBody(bytes: Uint8Array, start: number, end: number): Uint8Array<ArrayBuffer> {
  return bytes.subarray(start, end) as Uint8Array<ArrayBuffer>
}

/** Web path: start, pieces, commit, all same-origin under `/report/`. Only
 *  ever attempted when `canSend()` is true (the caller's responsibility —
 *  this function does not re-check). */
async function submitWeb(
  gzipBytes: Uint8Array,
  onProgress?: (progress: ReportProgress) => void,
): Promise<ReportSubmitResult> {
  const total = gzipBytes.byteLength
  const firstEnd = Math.min(PIECE_BYTES, total)

  const start = await fetchWithRetry('/report/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/gzip', 'Hew-Upload-Length': String(total) },
    body: pieceBody(gzipBytes, 0, firstEnd),
  })
  const startJson = await readJson(start)
  if (start.status !== 201) throw reportErrorFromResponse(start.status, startJson)
  const id = typeof startJson === 'object' && startJson !== null ? (startJson as { id?: unknown }).id : undefined
  const token = typeof startJson === 'object' && startJson !== null ? (startJson as { token?: unknown }).token : undefined
  if (typeof id !== 'string' || typeof token !== 'string') {
    throw new ReportError('status', 'malformed response from the report service')
  }
  onProgress?.({ sentBytes: firstEnd, totalBytes: total })

  const base = `/report/${encodeURIComponent(id)}`
  let offset = firstEnd
  for (let index = 1; offset < total; index++) {
    const end = Math.min(offset + PIECE_BYTES, total)
    const response = await fetchWithRetry(`${base}/${index}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/gzip', 'Hew-Upload-Token': token },
      body: pieceBody(gzipBytes, offset, end),
    })
    if (response.status !== 204) throw reportErrorFromResponse(response.status, await readJson(response))
    offset = end
    onProgress?.({ sentBytes: offset, totalBytes: total })
  }

  const commit = await fetchWithRetry(`${base}/commit`, {
    method: 'POST',
    headers: { 'Hew-Upload-Token': token },
  })
  if (commit.status === 201) return { id }
  throw reportErrorFromResponse(commit.status, await readJson(commit))
}

/** Submit a report — the gzip of the bundle `reportBundle.ts` builds
 *  (`send.gzip`) — to the intake service, reporting upload progress as it
 *  goes. Rejects with a `ReportError` naming the specific problem; callers
 *  should check `canSend()` first, since a build with nowhere to send has no
 *  reason to attempt this at all. */
export function submitReport(
  gzipBytes: Uint8Array,
  onProgress?: (progress: ReportProgress) => void,
): Promise<ReportSubmitResult> {
  return isTauri ? submitDesktop(gzipBytes, onProgress) : submitWeb(gzipBytes, onProgress)
}

File: shells/tauri/src-tauri/src/report_client.rs
//! The desktop's Help ▸ Report Bug submission client
//! (docs/design/report-bug.md §8 — the wire contract this must match
//! exactly).
//!
//! One Tauri command, `report_submit`, modeled on `relay_client.rs`'s
//! request plumbing: `reqwest` with the platform verifier stack (a homelab
//! CA trusted in the OS keychain is irrelevant here — the URL is fixed to
//! the Hew cloud — but the client is built the same way so there is only
//! one TLS story in this binary), typed errors, no new crates. Unlike the
//! relay, there is no per-user setting: every build mode sends to the same
//! `https://app.hew3d.com/report/` (design §5 — even a self-hosted desktop
//! reports bugs to Hew's developer, not the homelab). `HEW_REPORT_URL`
//! overrides that, but only in debug builds, so a release binary can never
//! be pointed at an attacker-controlled endpoint via the environment.
//!
//! The webview gzips the bundle, and the bytes ride the invoke as a raw body
//! (`tauri::ipc::Request`), the same reason `relay_put` does: an upload of up
//! to 90 MiB must not be serialized as tens of millions of numbers first.
//! This command then uploads them in pieces of `PIECE_BYTES` — start with the
//! first piece, each following piece in order, commit — because the intake
//! service gives each request only 10 ms of CPU (design §4). It emits
//! `report-progress` after each piece. Nothing here logs the body — not even
//! on error, since it can carry the user's own model and the files they
//! imported.

use std::time::Duration;

use serde::Serialize;
use tauri::ipc::InvokeBody;
use tauri::Emitter;

/// The intake service's fixed origin (docs/design/report-bug.md §8).
const REPORT_URL: &str = "https://app.hew3d.com/report/";

/// The upload piece size, one storage row on the intake service (§8).
const PIECE_BYTES: usize = 1_900_000;

/// One request carries at most one piece; two minutes covers that on a slow
/// uplink and still ends a stalled connection (design §8).
const REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
const CONNECT_TIMEOUT: Duration = Duration::from_secs(10);

// ---------------------------------------------------------------------------
// Errors — one kind per §8 status, plus the transport-level ones
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ReportErrorKind {
    /// 400 — not gzip, a head that doesn't parse, `format` ≠ 1, or
    /// description length out of range.
    Invalid,
    /// 413 — over the 90 MiB compressed cap.
    TooLarge,
    /// 429 — per-client rate limit.
    RateLimited,
    /// 507 — stored reports over the ceiling.
    Full,
    /// The request never completed at the transport level: DNS, refused,
    /// timeout, offline.
    Unreachable,
    /// TLS handshake failed.
    Tls,
    /// Any other unexpected HTTP status (403, 404, 409, 411, …) or a
    /// malformed response.
    Status,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportError {
    pub kind: ReportErrorKind,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<u16>,
}

impl ReportError {
    fn new(kind: ReportErrorKind, message: impl Into<String>) -> Self {
        Self {
            kind,
            message: message.into(),
            status: None,
        }
    }

    fn status(status: reqwest::StatusCode, message: impl Into<String>) -> Self {
        Self {
            kind: ReportErrorKind::Status,
            message: message.into(),
            status: Some(status.as_u16()),
        }
    }

    fn malformed() -> Self {
        Self::new(
            ReportErrorKind::Status,
            "malformed response from the report service",
        )
    }

    /// Maps one of §8's JSON error bodies (`{"error": "...", ...}`) plus the
    /// HTTP status onto a typed kind. Falls back to `Status` for anything
    /// that doesn't match the contract — a proxy's own error page, an
    /// unversioned server — rather than guessing.
    fn from_status(status: reqwest::StatusCode, body: &str) -> Self {
        #[derive(serde::Deserialize)]
        struct ErrorBody {
            error: String,
        }
        let code = serde_json::from_str::<ErrorBody>(body)
            .map(|b| b.error)
            .unwrap_or_default();
        match (status, code.as_str()) {
            (reqwest::StatusCode::BAD_REQUEST, "invalid") => Self::new(
                ReportErrorKind::Invalid,
                "the report was rejected as invalid",
            ),
            (reqwest::StatusCode::PAYLOAD_TOO_LARGE, "too-large") => {
                Self::new(ReportErrorKind::TooLarge, "the report is too large to send")
            }
            (reqwest::StatusCode::TOO_MANY_REQUESTS, "rate-limited") => Self::new(
                ReportErrorKind::RateLimited,
                "too many reports sent recently",
            ),
            (reqwest::StatusCode::INSUFFICIENT_STORAGE, "full") => Self::new(
                ReportErrorKind::Full,
                "the report service is temporarily full",
            ),
            _ if status.is_server_error() => Self::new(
                ReportErrorKind::Unreachable,
                "the report service is unavailable",
            ),
            _ => Self::status(status, format!("unexpected status {}", status.as_u16())),
        }
    }

    /// Classifies a transport-level failure — see `relay_client::RelayError::transport`,
    /// which this mirrors exactly (same reqwest/rustls stack, same reasoning
    /// for walking the source chain instead of matching reqwest's own
    /// Display, which embeds the request URL).
    fn transport(err: &reqwest::Error) -> Self {
        let mut text = String::new();
        let mut source = std::error::Error::source(err);
        while let Some(s) = source {
            text.push(' ');
            text.push_str(&s.to_string().to_ascii_lowercase());
            source = s.source();
        }
        if text.contains("certificate")
            || text.contains("unknownissuer")
            || text.contains("tls handshake")
            || text.contains("received fatal alert")
        {
            return Self::new(
                ReportErrorKind::Tls,
                "the server's certificate isn't trusted",
            );
        }
        if err.is_timeout() {
            return Self::new(ReportErrorKind::Unreachable, "timed out");
        }
        Self::new(
            ReportErrorKind::Unreachable,
            "could not reach the report service",
        )
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ReportSubmitResult {
    pub id: String,
}

/// How much of the compressed report has been uploaded, emitted as
/// `report-progress` after each piece.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct ReportProgress {
    sent_bytes: u64,
    total_bytes: u64,
}

/// Resolves the URL to post to: the fixed cloud origin in every release
/// build, or `HEW_REPORT_URL` when set AND this is a debug build (design
/// §8's "Clients" subsection) — for testing against `wrangler dev` without
/// ever giving a release binary an environment-controlled endpoint. Always
/// ends in `/`, so piece and commit paths join onto it.
fn report_url() -> String {
    #[cfg(debug_assertions)]
    if let Ok(url) = std::env::var("HEW_REPORT_URL") {
        if !url.is_empty() {
            return with_trailing_slash(url);
        }
    }
    REPORT_URL.to_owned()
}

fn with_trailing_slash(mut url: String) -> String {
    if !url.ends_with('/') {
        url.push('/');
    }
    url
}

/// The `[start, end)` byte ranges of the upload's pieces, in order; the
/// first carries the head the intake service validates.
fn piece_ranges(total: usize) -> Vec<(usize, usize)> {
    (0..total)
        .step_by(PIECE_BYTES)
        .map(|start| (start, (start + PIECE_BYTES).min(total)))
        .collect()
}

/// `HEW-` plus two groups of four Crockford base32 characters. The ID is
/// joined into request paths, so anything else is refused.
fn is_report_id(id: &str) -> bool {
    let bytes = id.as_bytes();
    bytes.len() == 13
        && id.starts_with("HEW-")
        && bytes[8] == b'-'
        && bytes
            .iter()
            .enumerate()
            .filter(|(i, _)| *i >= 4 && *i != 8)
            .all(|(_, b)| b.is_ascii_digit() || b.is_ascii_uppercase())
}

/// 32 random bytes, base64url without padding: 43 characters. The token
/// rides a header, so anything else is refused.
fn is_upload_token(token: &str) -> bool {
    token.len() == 43
        && token
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'-' || b == b'_')
}

fn build_client() -> reqwest::Client {
    // Same idempotent provider install as relay_client.rs — see its doc
    // comment for why this can't just rely on rustls's own default.
    let _ = rustls::crypto::ring::default_provider().install_default();
    reqwest::Client::builder()
        .connect_timeout(CONNECT_TIMEOUT)
        .timeout(REQUEST_TIMEOUT)
        .user_agent(concat!("Hew/", env!("CARGO_PKG_VERSION")))
        .build()
        .expect("reqwest client")
}

/// §8's retry rule: a network error or a 5xx is retried once. Every step
/// tolerates a repeat: the service accepts a retry of the last piece, and a
/// commit repeated with the upload's token answers 201 again.
async fn send_with_retry(
    build: impl Fn() -> reqwest::RequestBuilder,
) -> Result<reqwest::Response, ReportError> {
    let first = build().send().await;
    match first {
        Ok(response) if !response.status().is_server_error() => Ok(response),
        _ => build().send().await.map_err(|e| ReportError::transport(&e)),
    }
}

async fn error_from(response: reqwest::Response) -> ReportError {
    let status = response.status();
    let text = response.text().await.unwrap_or_default();
    ReportError::from_status(status, &text)
}

fn emit_progress(app: &tauri::AppHandle, sent: usize, total: usize) {
    // Progress is cosmetic; a failed emit never fails the upload.
    let _ = app.emit(
        "report-progress",
        ReportProgress {
            sent_bytes: sent as u64,
            total_bytes: total as u64,
        },
    );
}

/// Uploads the gzip-compressed report: start with the first piece, each
/// following piece in order, then commit. The webview passes the bytes as
/// the invoke's raw body (`invoke('report_submit', bytes)`), not a JSON
/// array, for the same reason `relay_put` does — see the module doc comment.
#[tauri::command]
pub async fn report_submit(
    app: tauri::AppHandle,
    request: tauri::ipc::Request<'_>,
) -> Result<ReportSubmitResult, ReportError> {
    let bytes: Vec<u8> = match request.body() {
        InvokeBody::Raw(bytes) => bytes.clone(),
        InvokeBody::Json(_) => {
            return Err(ReportError::new(
                ReportErrorKind::Invalid,
                "report_submit expects a raw byte body",
            ));
        }
    };
    if bytes.is_empty() {
        return Err(ReportError::new(
            ReportErrorKind::Invalid,
            "nothing to send",
        ));
    }

    let client = build_client();
    let base = report_url();
    let total = bytes.len();
    let pieces = piece_ranges(total);

    // Start: the first piece and the whole compressed length.
    let (first_start, first_end) = pieces[0];
    let response = send_with_retry(|| {
        client
            .post(&base)
            .header(reqwest::header::CONTENT_TYPE, "application/gzip")
            .header("Hew-Upload-Length", total.to_string())
            .body(bytes[first_start..first_end].to_vec())
    })
    .await?;
    if response.status() != reqwest::StatusCode::CREATED {
        return Err(error_from(response).await);
    }
    #[derive(serde::Deserialize)]
    struct StartBody {
        id: String,
        token: String,
    }
    let text = response
        .text()
        .await
        .map_err(|e| ReportError::transport(&e))?;
    let started: StartBody = serde_json::from_str(&text).map_err(|_| ReportError::malformed())?;
    if !is_report_id(&started.id) || !is_upload_token(&started.token) {
        return Err(ReportError::malformed());
    }
    emit_progress(&app, first_end, total);

    for (index, &(start, end)) in pieces.iter().enumerate().skip(1) {
        let url = format!("{base}{}/{index}", started.id);
        let response = send_with_retry(|| {
            client
                .put(&url)
                .header(reqwest::header::CONTENT_TYPE, "application/gzip")
                .header("Hew-Upload-Token", &started.token)
                .body(bytes[start..end].to_vec())
        })
        .await?;
        if response.status() != reqwest::StatusCode::NO_CONTENT {
            return Err(error_from(response).await);
        }
        emit_progress(&app, end, total);
    }

    let url = format!("{base}{}/commit", started.id);
    let response =
        send_with_retry(|| client.post(&url).header("Hew-Upload-Token", &started.token)).await?;
    if response.status() != reqwest::StatusCode::CREATED {
        return Err(error_from(response).await);
    }
    Ok(ReportSubmitResult { id: started.id })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_every_documented_status_to_its_kind() {
        let cases: &[(reqwest::StatusCode, &str, ReportErrorKind)] = &[
            (
                reqwest::StatusCode::BAD_REQUEST,
                r#"{"error":"invalid","message":"x"}"#,
                ReportErrorKind::Invalid,
            ),
            (
                reqwest::StatusCode::PAYLOAD_TOO_LARGE,
                r#"{"error":"too-large","maxBytes":94371840}"#,
                ReportErrorKind::TooLarge,
            ),
            (
                reqwest::StatusCode::TOO_MANY_REQUESTS,
                r#"{"error":"rate-limited"}"#,
                ReportErrorKind::RateLimited,
            ),
            (
                reqwest::StatusCode::INSUFFICIENT_STORAGE,
                r#"{"error":"full"}"#,
                ReportErrorKind::Full,
            ),
            (
                reqwest::StatusCode::INTERNAL_SERVER_ERROR,
                r#"{"error":"unavailable"}"#,
                ReportErrorKind::Unreachable,
            ),
            (
                reqwest::StatusCode::BAD_GATEWAY,
                r#"{"error":"unavailable"}"#,
                ReportErrorKind::Unreachable,
            ),
        ];
        for (status, body, expected) in cases {
            let err = ReportError::from_status(*status, body);
            assert!(
                matches!(&err.kind, k if std::mem::discriminant(k) == std::mem::discriminant(expected)),
                "{status}: got {:?}, expected {:?}",
                err.kind,
                expected
            );
        }
    }

    #[test]
    fn maps_piece_refusals_to_the_status_kind() {
        for (status, body) in [
            (reqwest::StatusCode::FORBIDDEN, r#"{"error":"forbidden"}"#),
            (reqwest::StatusCode::NOT_FOUND, r#"{"error":"not-found"}"#),
            (
                reqwest::StatusCode::CONFLICT,
                r#"{"error":"out-of-order","expected":3}"#,
            ),
            (
                reqwest::StatusCode::LENGTH_REQUIRED,
                r#"{"error":"length-required"}"#,
            ),
        ] {
            let err = ReportError::from_status(status, body);
            assert!(matches!(err.kind, ReportErrorKind::Status), "{status}");
            assert_eq!(err.status, Some(status.as_u16()));
        }
    }

    #[test]
    fn falls_back_to_status_kind_for_an_unrecognized_body() {
        let err = ReportError::from_status(reqwest::StatusCode::BAD_REQUEST, "not json at all");
        assert!(matches!(err.kind, ReportErrorKind::Status));
    }

    #[test]
    fn splits_an_upload_into_contiguous_pieces() {
        assert_eq!(piece_ranges(7), vec![(0, 7)]);
        assert_eq!(piece_ranges(PIECE_BYTES), vec![(0, PIECE_BYTES)]);
        assert_eq!(
            piece_ranges(PIECE_BYTES + 1),
            vec![(0, PIECE_BYTES), (PIECE_BYTES, PIECE_BYTES + 1)]
        );
        let cap = 90 * 1024 * 1024;
        let pieces = piece_ranges(cap);
        assert_eq!(pieces.len(), cap.div_ceil(PIECE_BYTES));
        assert!(pieces.windows(2).all(|w| w[0].1 == w[1].0));
        assert_eq!(pieces.last().map(|p| p.1), Some(cap));
        assert!(pieces.iter().all(|(s, e)| e - s <= PIECE_BYTES));
    }

    #[test]
    fn accepts_only_well_formed_ids_and_tokens() {
        assert!(is_report_id("HEW-7K3F-Q9XB"));
        assert!(!is_report_id("HEW-7K3F-Q9X"));
        assert!(!is_report_id("HEW-7K3F/Q9XB"));
        assert!(!is_report_id("HEW-../-Q9XB"));
        assert!(!is_report_id("hew-7k3f-q9xb"));

        assert!(is_upload_token(&"aZ09-_".repeat(8)[..43]));
        assert!(!is_upload_token(&"a".repeat(42)));
        assert!(!is_upload_token(&format!("{}\n", "a".repeat(42))));
    }

    #[test]
    fn report_url_defaults_to_the_cloud_origin_with_a_trailing_slash() {
        std::env::remove_var("HEW_REPORT_URL");
        assert_eq!(report_url(), REPORT_URL);
        assert_eq!(
            with_trailing_slash("http://127.0.0.1:8799/report".to_owned()),
            "http://127.0.0.1:8799/report/"
        );
    }
}
